python/lsst/dax/obscore/siav2.py

```diff
@@ -0,0 +1,523 @@
+            else:
+                times: list[astropy.time.Time | None] = []
+                for t in time_interval:
+                    if not math.isfinite(t):
+                        # Timespan uses None to indicate unbounded.
+                        times.append(None)
+                    else:
+                        times.append(astropy.time.Time(float(t), scale="utc", format="mjd"))
+                parsed["time"] = Timespan(times[0], times[1])
+        if calib:
+            parsed["calib"] = frozenset(calib)
+        return cls.model_validate(parsed)
+
+
+class SIAv2Handler:
+    """Process SIAv2 query parameters and convert into Butler query
+    constraints.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        Butler to query.
+    config : `ExporterConfig`
+        ObsCore configuration.
+    """
+
+    def __init__(self, butler: Butler, config: ExporterConfig):
+        self.butler = butler
+        self.config = config
+
+    def get_all_instruments(self) -> list[str]:
+        """Query butler for all known instruments.
+
+        Returns
+        -------
+        instruments : `list` [ `str` ]
+            All the instrument names known to this butler.
+        """
+        with self.butler._query() as query:
+            records = query.dimension_records("instrument")
+            return [rec.name for rec in records]
+
+    def get_band_information(self, instruments: list[str], band_interval: Interval) -> dict[str, Any]:
+        """Read all information from butler necessary to form a band query.
+
+        Parameters
+        ----------
+        instruments : `list` [ `str` ]
+            Instruments that could be involved in the band query.
+        band_interval : `Interval`
+            The band constraints.
+
+        Returns
+        -------
+        info : `dict` [ `str`, `typing.Any` ]
+            Dictionary that will be provided to `from_instrument_or_band`
+            when calculating the band and instrument queries.
+
+        Notes
+        -----
+        The base class implementation assumes the default dimension universe.
+        Each physical filter is checked against configuration to determine
+        whether it overlaps the requested band. If it does that filter is
+        stored in the returned dict with key ``filters`` where that dict maps
+        instrument names to a set of matching filters. Additionally, the
+        ``bands`` key is a set of bands associated with those filters
+        independent of instrument.
+        """
+        matching_filters = defaultdict(set)
+        matching_bands = set()
+        with self.butler._query() as query:
+            records = query.dimension_records("physical_filter")
+            if instruments:
+                instrs = ",".join(repr(ins) for ins in instruments)
+                records = records.where(f"instrument IN ({instrs})")
```

Comment:
I think this should also work:

    records = records.where("instrument in (INSTRUMENTS)", bind={"INSTRUMENTS": instruments})

Reply:
That's what I thought. I tried it but it definitely didn't work and the query system complained. Maybe I was doing it wrong so I'll try it again.

Reply:
Hmm, it should work in general, but maybe instrument have something special about it. If it complains we'll have to fix it.

Reply:
Butler was fixed to allow `IN` on governor dimensions lsst/daf_butler#1083